cmd/block-generator/runner/run.go

```diff
@@ -0,0 +1,260 @@
+	}
+	defer report.Close()
+
+	// Run for r.RunDuration
+	start := time.Now()
+	for time.Since(start) < r.RunDuration {
+		time.Sleep(r.RunDuration / 10)
+		if err := collector.Collect("indexer_daemon_import_time_sec"); err != nil {
+			return fmt.Errorf("problem collecting metrics: %w", err)
+		}
+	}
+	if err := collector.Collect("indexer_daemon_import_time_sec"); err != nil {
+		return fmt.Errorf("problem collecting metrics: %w", err)
+	}
+
+	// Collect results.
+	durationStr := fmt.Sprintf("test_duration_seconds:%d\n", uint64(r.RunDuration.Seconds()))
+	if _, err := report.WriteString(durationStr); err != nil {
+		return fmt.Errorf("unable to write duration metric: %w", err)
+	}
+
+	resp, err := http.Get(fmt.Sprintf("http://%s/report", generatorURL))
+	if err != nil {
+		return fmt.Errorf("the process failed to start properly, health endpoint query failed")
+	}
+	defer resp.Body.Close()
+	var generatorReport generator.Report
+	if err := json.NewDecoder(resp.Body).Decode(&generatorReport); err != nil {
+		return fmt.Errorf("problem decoding generator report: %w", err)
+	}
+	for metric, entry := range generatorReport {
+		// Skip this one
+		if metric == "genesis" {
+			continue
+		}
+		str := fmt.Sprintf("transaction_%s_total:%d\n", metric, entry.GenerationCount)
+		if _, err := report.WriteString(str); err != nil {
+			return fmt.Errorf("unable to write transaction_count metric: %w", err)
+		}
+	}
+
+	// Record a rate from one of the first data points.
+	if len(collector.Data) > 5 {
+		if err := recordDataToFile(collector.Data[2], "starting_block_import_duration_average_seconds", report); err != nil {
+			return err
+		}
+	}
+
+	// Also record the final one.
+	if err := recordDataToFile(collector.Data[len(collector.Data)-1], "final_import_duration_average_second", report); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// startGenerator starts the generator server.
+func startGenerator(configFile string, addr string) func() error {
+	// Start generator.
+	server, done := generator.StartServer(configFile, addr)
+
+	return func() error {
+		if err := server.Shutdown(context.Background()); err != nil {
+			return fmt.Errorf("failed during generator graceful shutdown: %w", err)
+		}
+
+		// Wait for graceful shutdown or crash.
+		select {
+		case <-done:
+			// continue
+			return nil
+		case <-time.After(10 * time.Second):
+			return fmt.Errorf("failed to gracefully shutdown generator")
+		}
+	}
+}
+
+// startIndexer resets the postgres database and executes the indexer binary. It performs some simple verification to
+// ensure that the service has started properly.
+func startIndexer(indexerBinary string, algodNet string, indexerNet string, postgresConnectionString string) (func() error, error) {
+	{
+		db, err := sql.Open("postgres", postgresConnectionString)
+		if err != nil {
+			return nil, fmt.Errorf("postgres connection string did not work: %w", err)
+		}
+		db.Exec(`DROP SCHEMA public CASCADE; CREATE SCHEMA public;`)
```

Comment:
Could you comment what this does?

Reply:
we should probably check the returned error

Reply:
Resets the DB. The returned error describes what it's trying to do now